sql/core/src/main/scala/org/apache/spark/sql/execution/Columnar.scala

```diff
@@ -458,6 +462,34 @@ case class RowToColumnarExec(child: SparkPlan) extends RowToColumnarTransition {
     // This avoids calling `schema` in the RDD closure, so that we don't need to include the entire
     // plan (this) in the closure.
     val localSchema = this.schema
+    if (enableArrowColumnVector) {
+      val maxRecordsPerBatch = SQLConf.get.arrowMaxRecordsPerBatch
+      val timeZoneId = SQLConf.get.sessionLocalTimeZone
+      return child.execute().mapPartitionsInternal { rowIterator =>
+        val context = TaskContext.get()
+        val allocator = ArrowUtils.getDefaultAllocator
+        val bytesIterator = ArrowConverters
+          .toBatchIterator(rowIterator, localSchema, maxRecordsPerBatch, timeZoneId, context)
```

Comment:
I'm not quite sure of your use case for this code, could you explain a bit more what this does for you, especially why serialize the record batch to a byte array?

Reply:
@BryanCutler , my goal here is to convert Iterator[InternalRow] to Iterator[ColumnarBatch] with ArrowColumnVectors as internal ColumnVector. The reason that I converted it to byte arrays first is inspired by @sunchao in previous comment, Quoted: 
>  instead of having a writable Arrow vector, currently other projects such as Iceberg and Spark RAPIDS choose to first construct the Arrow ValueVectors and then wrap them to Spark with ArrowColumnVector."

So I figured when I need to read from internalRow and write to arrow columnar format, there will be two ways to make that happen, 1) one is to read and write to ArrowRecordBatch and load as ColumnarBatch; 2) the other option is to use Arrow JAVA put APIs to write. 
And from @sunchao , I think option 1 would make more sense since Spark already has `ArrowConverters.toBatchIterator` function

Reply:
Same question here too. Why would we need another conversion step of Arrow format between `InternalRow` and `ColumnarBatch`? Are you looking for a Scala version of https://github.com/apache/spark/pull/34505?

Reply:
@HyukjinKwon , I looked into your PR and I am not sure what is the difference between your mapInArrow vs using pandas_udf, so I want to explain my PR vs pandas_udf here to avoid misunderstanding of your PR. 

Compared passing internalRows to pandas_udf, yes, the implementation here is very similiar while purpose is very different. My PR here is to support RowToColumnarExec sparkplan with arrow. By doing so, the following operators like sort, join, aggregate, project/filter can do computation based on Arrow format instead of using On/OffHeapColumnVector. 

And I think pandas_udf is used to do data transformation using pandas built-in computation or calling other thirdparty python package.

Not sure if I answered your question?

Reply:
Yeah, I am working on Scala version of it at SPARK-37229, and it will require to have a Spark physical plan that converts rows (or Columnar) to Arrow, and Arrow to Columnar. I am wondering if you can leverage this physical plan via using Spark extensions.

Reply:
@HyukjinKwon , yes, That is what we do, we have some implemented operators and plug into spark using Spark Extension configuration + columnarRules to replace original row physical plans, that is why we need RowToColumnarExec and ColumnarToRowExec to do InternalRow and ArrowColumnarBatch conversion. I think one difference with mapInArrow(scala version) is we want to leverage the built-in attribute supportColumnar to avoid code changes to spark application.

Reply:
Will you be able to add a separate `SparkPlan` that converts rows (or Columnar) to Arrow, and Arrow to Columnar instead of adding a configuration? Then I think I can leverage that physical plan in `mapInArrow`, and the change won't be invasive too.

For example,  you can add some physical plans like:

```scala
ColumnarToArrowExec
ArrowToColumnarExec
```

and then you use it via `SparkExtensions`. Does it make sense?

Reply:
@HyukjinKwon , `ArrowToColumnarExec`  is not needed because `ArrowColumnVector` is derived from `ColumnVector`, so current `ColumnarToRowExec` is capable of converting Arrow ColumnarBatch to InternalRow. 
And for `ColumnarToArrowExec`, there is some built-in rules [codes link](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/Columnar.scala#L512-L558) will add a  `ColumnarToRowExec` between two SparkPlans, if we use a new SparkPlan then we need to change above codes as well. So I am hoping we don't add these two new physical plans.
For mapInArrow case, how about we add a new parameter in RowToColumnarExec to tell it to use Arrow.

Reply:
@HyukjinKwon , I just realized I misunderstand what you meant, what do you mean by 'ColumnarToArrow` and `ArrowToColumnar`? so Arrow here is referring to Array Bytes of ArrowRecordBatch? And Columnar is ArrowColumnVector?

Reply:
Oh, okay. I guess there was a bit of miscommunication here.

I am more worrying about symmetry, and I am more talking about SQL physical plan level. I am thinking about having plans such as:

```scala
RowsToColumnarExec
ColumnarToRowsExec
ArrowToColumnarExec
ColumnarToArrowExec
```

What you want to do is I guess use Arrow-backed Columnar instances, right? You could, for example, pattern match on `RowsToColumnarExec(child)` and replace to something like `ColumnarToArrowExec(RowsToColumnarExec(child))` in Spark extensions.

I presume the overhead between `ColumnarBatch` and `ArrowBatch` is trivial.

Reply:
or you could replace `RowsToColumnarExec(child)` to `RowsToArrowExec(child)` as an example.

Reply:
yes, my question is why we need to add two extra physical plans?
If we make ArrowColumnVector as another backend of ColumnarBatch in `RowsToColumnarExec`, it can also be used by scala version of `mapInArrow`, doesn't it?
And another thing I am confused here is what is the `Arrow ` and `Columnar`  you are referring to in `ArrowToColumnarExec` and `ColumnarToArrowExec `context? To me, arrow is columnar, so are your talking about adding two Physical Plans to convert between `ArrowRecordBatch(type of RDD[Byte[Array]])` and `ColumnarBatch(type of RDD[ColumnarBatch])`

Reply:
Okay, so there are:
1. `ColumnarBatch`
2. `ColumnarBatch` backed by Arrow
3. `ArrowRecordBatch`

and what this PR does creates 2. from rows (instead of 1.). More specifically: `Rows -> 2` transition.

I was thinking something like:

`RowsToArrowExec` converts rows to 3. and `ArrowToColumnarExec` converts 2. More specifically `Rows -> 3 -> 2` transition.

I haven't taken a deeper look but I was thinking about that we should better have such physical plans and separations. I have been looking through related Arrow feature requests .. and my guts say it would be very useful to have these. At least I would like to reuse it.

Does it make sense?

Reply:
I see, I am not sure if you are familiar with [toArrowBatchRdd](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala#L3857-L3872) in Dataset,  it really looks like what `RowsToArrowExec` is.
And if your are looking for 

> converts 2. More specifically Rows -> 3 -> 2 transition.

I think that is what I am proposing in `RowToColumnarExec`, so if I think we can add a parameter in `RowToColumnarExec` to specify Arrow as backend, it already gets what we wanted.

After all, if you think adding `RowToArrowExec`  won't be duplicated to `Dataset.toArrowBatchRdd`, I am ok to add these two new physical plans, then do a combination of physical plan of converting Row -> 3 -> 2. I can do it in a seperate PR , but I still hope to make sure adding two new physical plans is a desired way. @cloud-fan @sunchao @BryanCutler

Reply:
(FWIW, I think we can deduplicate `toArrowBatchRdd` by using that new physical plan.)

Reply:
Cool, I'll work that. Will send out a PR later

Reply:
Sounds like this is working towards similar things. @xuechendi the reason I brought up the intermediate conversion to bytes is that it's an expensive step and not necessary if you are just converting  `ArrowRecordBatch` <-> `ColumnarBatch`. It's done in `ArrowConverters` specifically to send/read Arrow messages over a socket with Spark.

Reply:
Oh yeah. I think we shouldn't do the conversion step.

Reply:
@BryanCutler, are you saying converting between `ArrowRecordBatch` <-> `ColumnarBatch` is expensive? As far as I understand the codes, using Arrow `VectorSchemaRoot` to load `ArrowRecordBatch` just doing `ArrowBuf` reference_count increment then construct as a Arrow List[ValueVector], and this is the same method `ArrowConverters.fromBatchIterator` uses.
It will be great if you can help me to understand your expected way, thanks!
I would like to use any existing method in ArrowConverters or add some to it, the thing is the method you previously mentioned I replied before [`fromBatchIterator`](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowConverters.scala#L132-L136) is converting to RDD[InternalRow] not RDD[ColumnarBatch](This is why I didn't use it), but I can extract the shared part of the codes as a new function, and called by both `fromBatchIterator`(RDD[InternalRow]) and a new function, let's name it as `toColumnarBatchIterator`(RDD[ColumnarBatch])
Do you think it makes sense?

Reply:
@HyukjinKwon , what is the byte conversion step you are mentioning, is this for Iterator[InternalRow] to Iterator[ArrowRecordBatch]  or Iterator[ArrowRecordBatch] to Iterator[ColumnarBatch](From my understanding, this part is not doing heavy memory copy, as replied with some detail in last comment) ?